<a href="https://colab.research.google.com/github/danijel3/CroatianSpeech/blob/main/LM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Example of making an n-gram language model

The purpose of this document is to show the methodology used to produce the n-gram model used in other notebooks.

The tool used is [SRILM](http://www.speech.sri.com/projects/srilm/). This tool is very popular in ASR circles, although the license is slightly less conevinient than other tools.

Some other LM toolkits available:

* [MITLM](https://github.com/mitlm/mitlm) - very small, simple and easy to use, although a bit outdated and simplistic compared to others
- [KenLM](https://github.com/kpu/kenlm) - popular for its speed and efficieny, especially for large amounts of data
- [IRSTLM](https://github.com/irstlm-team/irstlm) - said to be more accurate in certain cases
- [PocoLM](https://github.com/danpovey/pocolm) - made by author(s) of Kaldi, specifically for their ASR - said to be better at fine-tuning to specific domains (see [motivation](https://github.com/danpovey/pocolm/blob/master/docs/motivation.md))

I will also recommend the [CMU Sphinx](https://cmusphinx.github.io/wiki/tutoriallm/) manual as a very approachable tutorial for someone starting out with this technique.

To use SRILM on your computer, you can either visit the website above and follow their instructions there, or if you are a user of Kaldi, you can use their excellent script in `./tools/extras/install_srilm.sh`.

For use in Colab, I have compiled a [version](https://github.com/danijel3/ASRforNLP/releases/tag/v1.0) of SRILM together with all other Kaldi tools that work in Colab. The script to unpack/install them is as follows (note, no GPU is needed for this notebook):

In [2]:
!wget https://github.com/danijel3/ASRforNLP/releases/download/v1.0/kaldi.tar.xz

!tar xvf kaldi.tar.xz -C / > /dev/null
%rm kaldi.tar.xz

!for f in $(find /opt/kaldi -name *.so*) ; do ln -sf $f /usr/local/lib/$(basename $f) ; done
!for f in $(find /opt/kaldi/src -not -name *.so* -type f -executable) ; do ln -s $f /usr/local/bin/$(basename $f) ; done
!for f in $(find /opt/kaldi/tools -not -name *.so* -type f -executable) ; do ln -s $f /usr/local/bin/$(basename $f) ; done

!ldconfig

--2021-12-24 21:57:50--  https://github.com/danijel3/ASRforNLP/releases/download/v1.0/kaldi.tar.xz
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/409506444/525a8238-abb3-4b8b-8282-12b094577f0e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211224%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211224T215750Z&X-Amz-Expires=300&X-Amz-Signature=29ccfd0c2a7f7bae67342d88c511920b0fee9c5859d656d662bcbc905eba56b6&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=409506444&response-content-disposition=attachment%3B%20filename%3Dkaldi.tar.xz&response-content-type=application%2Foctet-stream [following]
--2021-12-24 21:57:50--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/409506444/525a8238-abb3-4b8b-8282-12b094577f0e?X-Amz-Algorith

## LM basics

Let's take a simple "text corpus":

In [1]:
%%writefile simple.txt
ana ide u školu
ivo ide u školu
ivo ide kući

Writing simple.txt


Making a simple LM from a text file uses the `ngram-count` program (you can find a [complete manual](http://www.speech.sri.com/projects/srilm/manpages/ngram-count.1.html) online):

In [3]:
!ngram-count -text simple.txt -order 3 -wbdiscount -lm simple.arpa

The above command takes the following arguments:
* `text` - input text corpus - one sentence per line
* `order` - the "n" in n-gram, ie. trigram in this case
* `wbdiscount` - Witten-Bell discounting method - this one is not the best method, but better ones fail for very small text files
* `lm` - output language model file

The output file looks as follow:

In [4]:
%cat simple.arpa


\data\
ngram 1=8
ngram 2=9
ngram 3=3

\1-grams:
-0.7201593	</s>
-99	<s>	-0.2798407
-1.021189	ana	-0.2092596
-0.7201593	ide	-0.2798407
-0.8450981	ivo	-0.3853509
-1.021189	kući	-0.2092596
-0.8450981	u	-0.4101745
-0.8450981	školu	-0.3853509

\2-grams:
-0.69897	<s> ana
-0.39794	<s> ivo	0
-0.30103	ana ide
-0.69897	ide kući
-0.39794	ide u	0
-0.1760913	ivo ide
-0.30103	kući </s>
-0.1760913	u školu	0
-0.1760913	školu </s>

\3-grams:
-0.1760913	<s> ivo ide
-0.1760913	ide u školu
-0.1760913	u školu </s>

\end\


The ARPA-LM is a standrad format with a very long history. It starts with a hedaer beginning from a `/data/` token and counts of individual n-grams present in the file. Following that are lists of all the n-gram orders. You will note that a 3-gram model contains both unigrams and bigrams as well as trigrams, following the standard back-off methodology. 

Each n-gram line in the file contains two or three fields separated by tab characters (`\t`):
* probability of the n-gram in log (base 10) scale
* the name of the n-gram (ie. tokens/words separated by space)
* optional back-off weight, also log scale

Backing off allows us to acquire a probability value for a sequence of words, even if the n-gram isn't present in the model (eg. because it wasn't present in the training data). In such case, we can use a lower order n-gram but also have to apply the back-off weight (kind of as a form of punishment). The highest order n-grams (trigrams in our case) will never have any back-off weights. The algorithm for calculating the probability for any sequence of words is as follows:
1. look for the n-gram in the file and if you find it, return its value
2. if you dont find it, use the following formula:
\begin{equation}
P( word_N | word_{N-1}, word_{N-2}, ...., word_1 ) = \\
P( word_N | word_{N-1}, word_{N-2}, ...., word_2 ) \cdot \text{backoff-weight}(  word_{N-1} | word_{N-2}, ...., word_1 )
\end{equation}
3. if you can't find the n-gram of the lower order, then apply the rule recursively, all the way to unigrams (which should always exists, for words in vocabulary)
4. n-grams with now back-off weights are assumed to have value 1 (ie. 0 in log-scale)

Note, this assumes a colsed-vocabulary, that is the algorithm will fail for any word that is not inside the predefined vocabulary. A way around this is to add a special "UNK" token for any word not in the vocabulary. These are known as "out-of-vocabulary" words, or OOV. Modelling OOVs in n-grams does work, but obviously has its limitations (eg. one weight for any OOV for starters).

GOing back to the algorithm for a sec, lets takt a sequence `ana ide`. The probability of that would be:

\begin{equation}
P(ide|ana) = 10^{-0.30103} = 0.4999999950079739
\end{equation}

The probability of `ana ide kući`:

\begin{equation}
P(kući|ana,ide) = P(kući|ide)*bwt(ide|ana)=10^{(-0.69897+0)}=0.20000000199681048
\end{equation}

To verify these calculations, we can use the excellent python [ARPA](https://pypi.org/project/arpa/) library. This library contains no model preparation code, but can load any ARPA file and apply it to any n-gram or sentence. It is installed using pip:

In [7]:
!pip install arpa

Now we can compute the above values like so:

In [8]:
import arpa

lm=arpa.loadf('simple.arpa')[0]

print(lm.p('ana ide'))
print(lm.p('ana ide kući'))


0.4999999950079739
0.20000000199681048


## Larger LM

Let's create a mode realistic LM using a slightly bigger dataset:

In [9]:
!wget https://github.com/danijel3/CroatianSpeech/raw/main/ParlaMint-HR_S01.text.txt

--2021-12-24 22:30:11--  https://github.com/danijel3/CroatianSpeech/raw/main/ParlaMint-HR_S01.text.txt
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/danijel3/CroatianSpeech/main/ParlaMint-HR_S01.text.txt [following]
--2021-12-24 22:30:12--  https://raw.githubusercontent.com/danijel3/CroatianSpeech/main/ParlaMint-HR_S01.text.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4012131 (3.8M) [text/plain]
Saving to: ‘ParlaMint-HR_S01.text.txt’

ParlaMint-HR_S01.te 100%[===================>]   3.83M  --.-KB/s    in 0.07s   

2021-12-24 22:30:12 (57.8 MB/s) - ‘ParlaMint-HR_S01.text.txt’ saved [4012

This time we will use a slightly better discounting method (at least what people in ASR community found to be better in most cases). We will also use the interpolate function to tweak the weights a bit and add the UNK token for OOVs:

In [10]:
!ngram-count -text ParlaMint-HR_S01.text.txt -order 3 -kndiscount -interpolate -unk -lm hrvatski.arpa

When dealing with larger LMs, it is standard practice to compress the files. Since this is a text format, compression helps a lot. SRILM tools work fine with compressed files, as do many other toolkits, although they may have to have this feature added during compilation (eg. KenLM does). The PyTorch decoder does not support compressed LMs, afaik:

In [11]:
!gzip hrvatski.arpa

One thing that is standard practice with making these models is computing perplexity. This is a standard positive, but unconstrained measure, with lower values being better. To compute the value, we apply the model on an independent evaluation text, for example let's take this short output from our ASR:

In [12]:
%%writefile eval.txt
potpredsjedniče poštovane kolegice i kolege ovo je još jedan od zakona koji raspravljamo ovih dana ili ćemo raspravljati narednih dana koji su možda imali dobar motiv da budu upućeni hrvatskom saboru ali

Writing eval.txt


We use the `ngram` program to apply various computations on an already existsing LM:

In [13]:
!ngram -lm hrvatski.arpa.gz -unk -ppl eval.txt

file eval.txt: 1 sentences, 32 words, 0 OOVs
0 zeroprobs, logprob= -76.50246 ppl= 208.0924 ppl1= 245.8679


This command gave us to perplexity values: `ppl` and `ppl1`. The former is calculated only using the given words, and the latter is calculated by also applying the sentence boundary marks (ie. `<s>` and `</s>`). Whichever you use is totally up to you, but usually the choice doesn't matter much. 

The way we would use this information is if we change something in how the model is trained (eg. we use different parameters in the `ngram-count` command above, or we change the training data somehow), we can use PPL to compare the models and see which one is better.

Note that normally this program also calculates OOV, but for models that contain the UNK token, this value will unfortunately always be 0. This isn't a problem to count yourself, using a separate program/script, and is generally recommended as OOVs will often be a very big factor in WER.

Finally, we can use the `ngram` program to do fun stuff, like generate random sentences from the LM:

In [15]:
!ngram -lm hrvatski.arpa.gz -unk -gen 10

Kakve Marasu, bili danas ako ste vi tako jednom tretman gdje rješenja bilo njihova javne dužnosti.
Hvala potpredsjedniče Hrvatskog sabora.
Ja ću i ruralnih „Isušiti močvaru“. predsjednik. ima ovrhu gospodarstva, radimo ova dva dana samo što će zakon.
Idući sustava, situaciju reforme i HSU-a, smije koji to znači stavak 1., prema toj identificirate jako svinjogojstvu, pomorske snage i godinu dana modela i ostvariti iskreno sumnjamo. počešali. klubova to ne riješe dovede u BiH, da onaj tko je bilo.
Zahvaljujem zastupniku smo u Europskom novine nije.
i to je da on nije uvjeravam stranih onda se radi.
To su vrlo ne ostvarena i niste mogli imati ogroman srpskom dajte pred proširen katalog kaznenih djela uvaženog grada, facebooku korištenje europskih dakle u sadašnjost.
Pa stimulativno djelovala. stanovništva a i osobe se da su manje i sustav koji će uslijediti ili tako unaprijedite i čitav niz pitanja ako se stalo nepromišljeno nelogičnosti kod nas izabrale.
Poštovani predsjedniče Hrvatskog 

## Final words

This is where this small document ends for now. The above set of random sentences shows pretty well what the limitations of this method is. To fully control this technique, several more topics are recommended for in-depth study:

* various discounting and interpolation methods and their meaning for the final result
* managing size and complexity of models, eg. by pruning
* considering the combination of char-level and LM-based decoding in E2E systems (the decoder has a weight which allows to consider either option to a certaing degree)
* dealing with [large-scale LMs](https://cmusphinx.github.io/wiki/tutoriallmadvanced/), model mixing and domain adaptation
* other advanced n-gram techniques like: skip-grams, class-based LMs, factored LMs, hidden event LMs, sub-word modeling, etc...